File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.go
// time o(n), space o(1)
func maxProfitS(prices []int) int {
    maxprofit := 0
    
    // simplified
    minprice := prices[0]
    for i := 1; i < len(prices); i++ {
        minprice = min(minprice, prices[i])
        maxprofit = max(maxprofit, prices[i] - minprice)
    }
    
    return maxprofit
}

// go doesn't have min/max method for integers
func max(a,b int) int{
    if a > b {
        return a
    }
    return b
}
func min(a,b int) int{
    if a > b{
        return b
    }
    return a
}


func maxProfit(prices []int) int {
    ////  init left, right, maxprofit, slice length
    left := 0
    right := 1
    maxprofit := 0
    length := len(prices)
    
    if length == 0{
        return maxprofit
    }
    
    for right < length{
        if prices[left] < prices[right]{
            // check max profit
            maxprofit = max(maxprofit, prices[right] - prices[left])
        }else{
            // since price[right] has lower selling price,
            // update left pointer to have right pointer index value
            left = right
        }
        // keep moving right pointer
        right += 1
    }
    
    return maxprofit
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int l = 0;
        int r = 1;
        int max = 0;

        while(l < prices.size()){
            int lp = prices[l];
            int rp = prices[r];
            if(lp > rp){
                l++;
                r++;
                if(r >= prices.size()){
                    r = prices.size();
                }
                continue;
            }
            if(lp < rp){
                max = std::max(max, rp - lp);
                r++;
                if(r >= prices.size()){
                    r = prices.size();
                }
            }
            if(prices[l] > prices[r]){
                l++;
            }
        }

        return max;
    }
};